File: leetcode/73_MatrixSetZeros/lc73.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        vector<vector<int>> index;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == 0) {
                    if (index.size() == 0) {
                        vector<int> new_element;
                        new_element.push_back(i);
                        new_element.push_back(j);
                        index.push_back(new_element);
                    } else {
                        int added = 0;
                        for (int k = 0; k < index.size(); k++) {
                            if (index[k][0] == i) {
                                added = 1;
                                index[k].push_back(j);
                            }
                        }
                        if (added == 0) {
                            vector<int> new_element;
                            new_element.push_back(i);
                            new_element.push_back(j);
                            index.push_back(new_element);
                        }
                    }
                }
            }
        }
        if (index.size() == 0) return;
        for (int i = 0; i < index.size(); i++) {
            int row = index[i][0];
            for (int j = 0; j < n; j++) {
                matrix[row][j] = 0;
            }
            for (int j = 1; j < index[i].size(); j++) {
                int col = index[i][j];
                for (int k = 0; k < m; k++) {
                    matrix[k][col] = 0;
                }
            }
        }
    }
};

File: leetcode/73_MatrixSetZeros/lc73.py
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        m, n = len(matrix), len(matrix[0])
        flag = False

        for i in range(m):
            if not matrix[i][0]:
                flag = True
            for j in range(1, n):
                if matrix[i][j] == 0:
                    matrix[0][j] = 0
                    matrix[i][0] = 0
        
        for i in range(m):
            for j in range(1, n):
                if matrix[m-i-1][0] == 0 or matrix[0][j] == 0:
                    matrix[m-i-1][j] = 0
            if flag:
                matrix[m-i-1][0] = 0
